src/Internal/DependencyManagement/ExtendedContainer.php

```diff
@@ -43,22 +51,21 @@ class ExtendedContainer extends \League\Container\Container {
 	 * @throws ContainerException Invalid parameters.
 	 */
 	public function add( string $class_name, $concrete = null, bool $shared = null ) : DefinitionInterface {
-		if ( ! $this->class_is_in_root_namespace( $class_name ) && ! in_array( $class_name, $this->registration_whitelist, true ) ) {
-			throw new ContainerException( "Can't use the container to register '$class_name', only objects in the " . Container::WOOCOMMERCE_ROOT_NAMESPACE . ' namespace are allowed for registration.' );
+		if ( ! $this->is_class_allowed( $class_name ) ) {
+			throw new ContainerException( "You cannot add '$class_name', only classes in the {$this->woocommerce_namespace} namespace are allowed." );
 		}
 
-		return parent::add( $class_name, $concrete, $shared );
-	}
+		$concrete_class = $this->get_class_from_concrete( $concrete );
```

Comment:
I should note that this isn't entirely comprehensive because they can set a closure to a concrete that returns an invalid class, but given how unlikely that is, I think this should be good enough. I considered adding some error mechanism to `Definition` for the callable case but I felt like that might be overkill?

Reply:
Using an obscure concrete like that would be akin to using reflection to invoke a private method: not something that any of us would approve in a PR except in very very special cases. So indeed, I think extra error checking isn't necessary.